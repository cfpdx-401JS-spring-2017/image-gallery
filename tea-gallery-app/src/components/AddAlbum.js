import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddAlbum extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
    };
    
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({ [name]: value });
  }

  render() {
    return (
      <form className="add-item" onSubmit={e => {
        e.preventDefault();
        this.props.onAddAlbum(this.state);
      }}>
        <label>Add an album<input placeholder='album name' name='name' onChange={this.handleChange}></input></label>
        <button type="submit">Add</button>
      </form>
    );
  }
}

AddAlbum.PropTypes = {
  onAddAlbum: PropTypes.func.isRequired
};

export default AddAlbum;