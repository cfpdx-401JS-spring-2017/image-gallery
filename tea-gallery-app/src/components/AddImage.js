import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddImage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      title: '',
      description: '',
      url: ''
    };
    
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({ [name]: value });
  }

  render() {
    return (
      <form className="add-item" onSubmit={e => {
        e.preventDefault();
        this.props.onAddImage(this.state);
      }}>
        <label>Image name<input name='title' onChange={this.handleChange}></input></label><br/>
        <label>Description<input name='description' onChange={this.handleChange}></input></label><br/>
        <label>URL<input name='url' onChange={this.handleChange}></input></label><br/>
        <button type="submit">Add</button>
      </form>
    );
  }
}

AddImage.propTypes = {
  onAddImage: PropTypes.func,
};

export default AddImage;