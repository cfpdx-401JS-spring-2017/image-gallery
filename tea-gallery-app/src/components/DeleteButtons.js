import React from 'react';

export function DeleteImageButton(props) {
  return (
    <button onClick={props.onDeleteImage}>Delete</button>
  );
}

export function DeleteAlbumButton(props) {
  return (
    <button onClick={props.onDeleteAlbum}>Delete album</button>
  );
}