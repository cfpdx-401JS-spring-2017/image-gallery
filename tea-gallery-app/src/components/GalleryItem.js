import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DeleteImageButton } from './DeleteButtons';

class GalleryItem extends Component {

  static PropTypes = {
    images: PropTypes.array.required,
    onDeleteImage: PropTypes.func.required
  };

  constructor(props) {
    super(props);

    this.state = {
      index: 0
    };
  }

  changeImage(increment) {
    const images = this.props.images;
    let index = this.state.index;
    const image = images[index];

    index += increment;
    if (index === images.length) index = 0;
    if (index === -1 || isNaN(index)) index = images.length - 1;

    this.setState({ index: index });
    this.setState({ images: image });
  }

  render() {
    const { images } = this.props;
    return (
      <div className="gallery-img-wrapper">
        {images.length > 0 &&
          <div>
            <img className="gallery-img" src={images[this.state.index].url} alt="tea"></img>
            <button className="back-button" onClick={() => this.changeImage(-1)}>Back</button>
            <button className="next-button" onClick={() => this.changeImage(+1)}>Next</button>
            <DeleteImageButton
              images={images}
              onDeleteImage={() => this.props.onDeleteImage(images[this.state.index]._id)}
            />
          </div>
        }
        {images.length < 1 && <h2>Album is empty. Add images below</h2>}
      </div>
    );
  }
}

export default GalleryItem;