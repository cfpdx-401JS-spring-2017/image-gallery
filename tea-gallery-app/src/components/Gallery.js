import React from 'react';
import PropTypes from 'prop-types';
import GalleryItem from './GalleryItem';

Gallery.PropTypes = {
  images: PropTypes.array,
  onDeleteImage: PropTypes.func.required
};

export default function Gallery(props) {
  return (
    <GalleryItem
      images={props.images}
      onDeleteImage={props.onDeleteImage} 
    />
  );
}