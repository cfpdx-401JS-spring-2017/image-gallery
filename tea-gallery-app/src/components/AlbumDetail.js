import React, { Component } from 'react';
import { Link, Route, Switch, Redirect } from 'react-router-dom';
import AddImage from './AddImage';
import albumApi from '../services/albumApi';
import Thumbnail from './Thumbnail';
import List from './List';
import Gallery from './Gallery';
import { DeleteAlbumButton } from './DeleteButtons';

const views = [Thumbnail, List, Gallery];

class AlbumDetail extends Component {
  constructor(props) {
    super(props);

    this.state = {
      album: null
    };

    this.onAddImage = this.onAddImage.bind(this);
    this.onDeleteImage = this.onDeleteImage.bind(this);
  }

  getAlbumWrapper(albumId) {
    albumApi.getAlbum(albumId)
      .then(album => this.setState({ album }));
  }

  componentDidMount() {
    const { albumId } = this.props.match.params;
    this.getAlbumWrapper(albumId);
  }

  componentWillReceiveProps({ match }) {
    this.getAlbumWrapper(match.params.albumId);
  }

  onDeleteImage(imageId) {
    albumApi.deleteImage(imageId, this.state.album._id)
      .then(() => {
        const images = this.state.album.images.slice();
        const index = images.findIndex(img => img._id === imageId);
        images.splice(index, 1);
         
        const updatedAlbum = Object.assign({}, this.state.album );
        updatedAlbum.images = [...images];
        this.setState({ album: updatedAlbum });
      });
  }

  onAddImage(image) {
    albumApi.addImage(image, this.state.album._id)
      .then(image => {
        const updatedAlbum = Object.assign({}, this.state.album );
        updatedAlbum.images = [...this.state.album.images, image];

        this.setState({
          album: updatedAlbum
        });
      });
  }

  render() {
    const { match } = this.props;
    if (!this.state.album) return <br/>;
    return (
      <div className="wrapper">
        <div className="view-button-wrapper">
          {views.map(View =>
            <Link key={View.name} to={`${match.url}/${View.name}`}>
              <button key={View.name}>
                {View.name}
              </button>
            </Link>)
          }
        </div>
        <Switch>
          <Route path={`${match.url}/:view`}
            render={(props) => {
              const View = views.find(view => view.name === props.match.params.view);
              return <View
                onDeleteImage={this.onDeleteImage}
                images={this.state.album.images}
              />;
            }}
          />
          <Redirect from={match.url} to={`${match.url}/Thumbnail`} />
        </Switch>
        <DeleteAlbumButton
          album={this.state.album}
          onDeleteAlbum={() => this.props.onDeleteAlbum(this.state.album._id)}
        />
        <AddImage onAddImage={this.onAddImage} />
      </div >
    );
  }
}

export default AlbumDetail;

