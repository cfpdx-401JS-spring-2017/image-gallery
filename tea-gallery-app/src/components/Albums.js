import React, { Component } from 'react';
import { Route, Link } from 'react-router-dom';
import albumApi from '../data';
import AddAlbum from './AddAlbum';
import AlbumDetail from './AlbumDetail';

class Albums extends Component {

  constructor(props) {
    super(props);

    this.state = {
      albums: [],
      showAddAlbum: true,
      showAddImage: false
    };

    this.onAddAlbum = this.onAddAlbum.bind(this);
    this.onDeleteAlbum = this.onDeleteAlbum.bind(this);
  }

  componentDidMount() {
    albumApi.get()
      .then(albums => {
        this.setState({
          albums
        });
      });
  }

  onDeleteAlbum(id) {
    albumApi.deleteAlbum(id)
      .then(() => {
        const albums = this.state.albums;
        const index = albums.findIndex(album => album._id === id);
        albums.splice(index, 1);
        this.setState({ albums });
      });
  }

  onAddAlbum(album) {
    album.images = [];
    albumApi.AddAlbum(album)
      .then(album => {
        this.setState({
          albums: [...this.state.albums, album]
        });
      });
  }

  render() {
    const { match } = this.props;
    const { albums } = this.state;
    return (
      <div>
        <div className="albums-view">
          <h3>Albums</h3>
          {albums.map(album =>
            <Link key={album.name} to={`/albums/${album._id}`}>
              <button
              key={album.name}
                onClick={() => this.setState({ showAdd: false })}
              >
                {album.name}
              </button>
            </Link>)}
          {this.state.showAddAlbum && <AddAlbum onAddAlbum={this.onAddAlbum} />}
        </div>
        <Route path={`${match.url}/:albumId`}
          render={({ match }) => {
            return <AlbumDetail
              match={match}
              onDeleteAlbum={this.onDeleteAlbum} />;
          }} />
      </div>
    );
  }
}

export default Albums;
