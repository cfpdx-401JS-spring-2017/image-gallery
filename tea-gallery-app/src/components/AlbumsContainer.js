import React, { Component } from 'react';
import albumApi from '../services/albumApi';
import Albums from './Albums';

class AlbumsContainer extends Component {

  constructor(props) {
    super(props);

    this.state = {
      albums: []
    };

    this.onAddAlbum = this.onAddAlbum.bind(this);
    this.onDeleteAlbum = this.onDeleteAlbum.bind(this);
  }

  componentDidMount() {
    albumApi.getAll()
      .then(albums => {
        this.setState({
          albums
        });
      });
  }

  onDeleteAlbum(id) {
    albumApi.deleteAlbum(id)
      .then(() => {
        const albums = this.state.albums.slice();
        const index = albums.findIndex(album => album._id === id);
        albums.splice(index, 1);
        this.props.history.push('/albums');
        this.setState({ albums });
      });
  }

  onAddAlbum(album) {
    album.images = [];
    albumApi.addAlbum(album)
      .then(album => {
        this.setState({
          albums: [...this.state.albums, album]
        });
      });
  }

  render() {
    return (
      <Albums
        albums={this.state.albums}
        onAddAlbum={this.onAddAlbum}
        onDeleteAlbum={this.onDeleteAlbum}
      />
    );
  }
}

export default AlbumsContainer;
