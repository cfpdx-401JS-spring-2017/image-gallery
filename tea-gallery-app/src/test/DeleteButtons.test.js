import React from 'react';
import renderer from 'react-test-renderer';
import { DeleteAlbumButton, DeleteImageButton } from '../components/DeleteButtons';

const images = [
  {
    title: 'tea leaves in teapot',
    description: 'what nice leaves',
    url: '/../public/img/teainpot.jpg'
  },
  {
    title: 'Tetsubin over charcoal',
    description: 'what a nice tetsubin',
    url: '/../public/img/tetsubin-charcoal.jpg'
  },
  {
    title: 'Teapot display',
    description: 'what a nice teapot display',
    url: '/../public/img/teapot-display.jpg'
  }
];

describe('DeleteButtons', () => {
  it('renders the DeleteAlbumButton component', () => {
    const component = renderer.create(<DeleteAlbumButton images={images}/>);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders the DeleteImageButton component', () => {
    const component = renderer.create(<DeleteImageButton images={images}/>);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

});