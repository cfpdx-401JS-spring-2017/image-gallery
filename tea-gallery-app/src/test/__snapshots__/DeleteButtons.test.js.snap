// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteButtons renders the DeleteAlbumButton component 1`] = `
<button
  onClick={undefined}
>
  Delete album
</button>
`;

exports[`DeleteButtons renders the DeleteImageButton component 1`] = `
<button
  onClick={undefined}
>
  Delete
</button>
`;
